### PR TITLE
Add stubs for more unimplemented reflect methods

### PR DESCRIPTION
**Pull request by kyleconroy**
I'm current working on solving #447 by getting https://github.com/protocolbuffers/protobuf-go to compile with TinyGo. In combination with https://github.com/planetscale/vtprotobuf, I think there's a chance I get the code generated by protoc to compile and run without any changes.

This is my first attempt to "implement" some of the missing methods in the reflect package. This is my first time contributing to TinyGo, so I have no idea if I'm doing things correctly. Please let me know!

I've included some code copied directly from the Go source code. Is this allowed? How do I mark that code?

### CODE REVIEW

File: src/reflect/type.go [hunk 1]
```diff
@@ -126,6 +126,23 @@ func (k Kind) basicType() rawType {
...
+	// method name. It is empty for upper case (exported) method names.
+	// The combination of PkgPath and Name uniquely identifies a method
+	// in a method set.
+	// See https://golang.org/ref/spec#Uniqueness_of_identifiers
+	PkgPath string
+
+	Type  Type  // method type
+	Func  Value // func with receiver as first argument
+	Index int   // index for Type.Method
+}
```
Comment (kyleconroy) [inline src/reflect/type.go:144]: Copied from Go source

File: src/reflect/value.go [hunk 2]
```diff
@@ -788,10 +788,14 @@ type stringHeader struct {
...
 	Method string
+	Kind   Kind
 }
 
 func (e *ValueError) Error() string {
-	return "reflect: call of reflect.Value." + e.Method + " on invalid type"
+	if e.Kind == 0 {
+		return "reflect: call of " + e.Method + " on zero Value"
+	}
+	return "reflect: call of " + e.Method + " on " + e.Kind.String() + " Value"
```
Comment (kyleconroy) [inline src/reflect/value.go:798]: Copied from Go source

File: src/reflect/value.go [hunk 3]
```diff
@@ -187,7 +187,7 @@ func (v Value) Bool() bool {
 			return uintptr(v.value) != 0
 		}
 	default:
-		panic(&ValueError{"Bool"})
+			panic(&ValueError{Method: "Bool"})
```
Comment (kyleconroy) [inline src/reflect/value.go]: Not sure how all the spacing got messed up here. Happy to fix this up if this is something you'd like to merge.

File: src/reflect/value.go [hunk 4]
```diff
@@ -788,10 +788,14 @@ type stringHeader struct {
 
 type ValueError struct {
 	Method string
+	Kind   Kind
```
Comment (kyleconroy) [inline src/reflect/value.go:791]: Updated to match the ValueError from Go

File: src/reflect/type.go [hunk 5]
```diff
@@ -691,6 +708,23 @@ func (t rawType) Key() Type {
...
+	panic("unimplemented: (reflect.Type).Out()")
+}
+
+func (t rawType) MethodByName(name string) (Method, bool) {
+	panic("unimplemented: (reflect.Type).MethodByName()")
+}
+
+func (t rawType) PkgPath() string {
+	// panic("unimplemented: (reflect.Type).PkgPath()")
+	return "<unimplemented>"
```
Comment (kyleconroy) [inline src/reflect/type.go]: I need this to return a string for protocol buffers to work, but I'm happy to put the panic back in so the code at least compiles.

**Comment by dkegel-fastly** [comment]
Copying source from upstream is fine, just make sure the file has copyright headers.

I haven't looked at reflect, so I'm not the person to comment, but:
Please include a test for the particular thing you're fixing.  (e.g. a test that uses DeepEquals to compare two things that have the type you're improving support for).

**Comment by jrauschenbusch** [comment]
Do you also plan to implement reflect methods required by gob?

https://tinygo.org/docs/reference/lang-support/stdlib/#encodinggob

**Comment by kyleconroy** [comment]
> Please include a test for the particular thing you're fixing.

Will do, not sure how unimplemented methods are currently tested.

> Do you also plan to implement reflect methods required by gob?

I do not. If you look at my change, I'm not implementing any methods, only adding unimplemented stubs so that the default code generated by protoc compiles. That code will still fail at runtime.

**Comment by dkegel-fastly** [comment]
Oh, I didn't realize you were just adding stubs.  Changed title to reflect what you said.

You can add smoke tests that only verify compilation, and I've done that before, but I'm not sure it's worth it in this case.

**Comment by kyleconroy** [comment]
@dkegel-fastly Would it be possible to approve the 3 workflows for this PR? I'd like to know if my change passes all the tests.

**Comment by dkegel-fastly** [comment]
that is kind of a surprising failure...

**Comment by kyleconroy** [comment]
@dkegel-fastly I've rebased on the latest dev, can you approve the tests again?

**Comment by kyleconroy** [comment]
Great, looks like all the tests are passing. Anything else I need to do to get this merged in?

**Comment by dkegel-fastly** [comment]
Check with @dgryski, he may have suggestions.  In slack, he wrote "I wonder if stubbing more of reflect would allow text/template to build and be used for simple things (that don't use the bits we faked)". 

I just checked tinygo building example_test.go from upstream's text/template against dev and against your branch.  Resulting errors (ones you got rid of are marked with a dash):
```
  SendDir not declared by package reflect
  etyp.FieldByName undefined (type reflect.Type has no field or method FieldByName)
  item.Slice3 undefined (type reflect.Value has no field or method Slice3)
- ptr.MethodByName undefined (type reflect.Value has no field or method MethodByName)
  receiver.Type().FieldByName undefined (type reflect.Type has no field or method FieldByName)
- typ.In undefined (type reflect.Type has no field or method In)
  typ.IsVariadic undefined (type reflect.Type has no field or method IsVariadic)
  typ.NumIn undefined (type reflect.Type has no field or method NumIn)
  typ.NumOut undefined (type reflect.Type has no field or method NumOut)
- typ.Out undefined (type reflect.Type has no field or method Out)
  v.Type().NumOut undefined (type reflect.Type has no field or method NumOut)
  val.Recv undefined (type reflect.Value has no field or method Recv)
  val.Type().ChanDir undefined (type reflect.Type has no field or method ChanDir)
```

Whew.  Well, getting text/template to build its example_test.go means stubbing another ten methods.  That's a lot.

Is there some open source package that your patch rescues?  That might help.

**Comment by kyleconroy** [comment]
> Is there some open source package that your patch rescues? That might help.

Yes! This patch, combined with https://github.com/planetscale/vtprotobuf, allows tinygo programs to use Protocol Buffers*. I'm also happy to add stubs for the rest of those methods if it will mean this gets merged.

* There is still one small change you have to make to the generated code, but I'd commit to fixing that once this is merged.

**Comment by dkegel-fastly** [comment]
Closing in favor of #2624.

**Pull request closed without merging**